File: swift/sql_injection/ProductQueries.swift
// ProductQueries.swift
// Shopist – Product catalogue data access layer
// ⚠️  DEMO FILE – Intentionally vulnerable for Datadog Code Security demo
// DO NOT USE IN PRODUCTION

import Foundation
import SQLite3

class ProductQueries {

    var db: OpaquePointer?

    init(databasePath: String) {
        if sqlite3_open(databasePath, &db) != SQLITE_OK {
            print("Failed to open database at \(databasePath)")
        }
    }

    deinit {
        sqlite3_close(db)
    }

    // VULN 1: SQL Injection via string interpolation in product search
    // The `searchTerm` value comes directly from a query parameter and is
    // interpolated without escaping or binding.
    func searchProducts(searchTerm: String) -> [[String: Any]] {
        var products: [[String: Any]] = []

        // Attacker input: ' UNION SELECT id,username,password_hash,4,5 FROM users--
        let query = "SELECT id, name, description, price, stock FROM products WHERE name LIKE '%\(searchTerm)%' OR description LIKE '%\(searchTerm)%'"
        var statement: OpaquePointer?

        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            while sqlite3_step(statement) == SQLITE_ROW {
                let id = sqlite3_column_int(statement, 0)
                let name = String(cString: sqlite3_column_text(statement, 1))
                let description = String(cString: sqlite3_column_text(statement, 2))
                let price = sqlite3_column_double(statement, 3)
                let stock = sqlite3_column_int(statement, 4)
                products.append(["id": id, "name": name, "description": description, "price": price, "stock": stock])
            }
        }
        sqlite3_finalize(statement)
        return products
    }

    // VULN 2: SQL Injection via string concatenation for price-range filter
    // `minPrice` and `maxPrice` arrive as raw strings from the request; they are
    // concatenated directly into the SQL without validation or binding.
    func filterProductsByPrice(minPrice: String, maxPrice: String) -> [[String: Any]] {
        var products: [[String: Any]] = []

        // Attacker input for minPrice: 0 OR 1=1--
        let sql = "SELECT id, name, price, category_id FROM products WHERE price >= " + minPrice +
                  " AND price <= " + maxPrice + " AND is_active = 1"
        var statement: OpaquePointer?

        if sqlite3_prepare_v2(db, sql, -1, &statement, nil) == SQLITE_OK {
            while sqlite3_step(statement) == SQLITE_ROW {
                let id = sqlite3_column_int(statement, 0)
                let name = String(cString: sqlite3_column_text(statement, 1))
                let price = sqlite3_column_double(statement, 2)
                let categoryId = sqlite3_column_int(statement, 3)
                products.append(["id": id, "name": name, "price": price, "categoryId": categoryId])
            }
        }
        sqlite3_finalize(statement)
        return products
    }

    // VULN 3: SQL Injection via String(format:) in category query
    // The category slug is a user-supplied string passed as a %@ format argument –
    // String(format:) performs plain text substitution, not parameterisation.
    func getProductsByCategory(categorySlug: String) -> [[String: Any]] {
        var products: [[String: Any]] = []

        // Attacker input: electronics' OR '1'='1
        let query = String(format: "SELECT p.id, p.name, p.price, p.image_url FROM products p JOIN categories c ON p.category_id = c.id WHERE c.slug = '%@' AND p.is_active = 1", categorySlug)
        var statement: OpaquePointer?

        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            while sqlite3_step(statement) == SQLITE_ROW {
                let id = sqlite3_column_int(statement, 0)
                let name = String(cString: sqlite3_column_text(statement, 1))
                let price = sqlite3_column_double(statement, 2)
                let imageUrl = String(cString: sqlite3_column_text(statement, 3))
                products.append(["id": id, "name": name, "price": price, "imageUrl": imageUrl])
            }
        }
        sqlite3_finalize(statement)
        return products
    }
}

File: php/sql_injection/product_queries.php
<?php
/**
 * Shopist - Product Query Functions
 * DEMO FILE: Intentionally vulnerable for Datadog Code Security SAST demo.
 * DO NOT USE IN PRODUCTION.
 */

$conn = mysqli_connect("localhost", "shopist_user", "shopist_pass", "shopist_db");

// VULN 1: String concatenation in product search query using $_GET (SQL Injection)
function searchProducts($conn) {
    // $_GET['q'] inserted directly into the query without sanitization
    $searchTerm = $_GET['q'];
    $query = "SELECT id, name, price, stock FROM products WHERE name LIKE '%" . $searchTerm . "%' OR description LIKE '%" . $searchTerm . "%'";
    $result = mysqli_query($conn, $query);
    $products = [];
    while ($row = mysqli_fetch_assoc($result)) {
        $products[] = $row;
    }
    return $products;
}

// VULN 2: $_GET parameter used directly in SQL WHERE clause (SQL Injection)
function getProductById($conn) {
    // Product ID taken from query string and placed directly into SQL
    $productId = $_GET['id'];
    $result = mysqli_query($conn, "SELECT * FROM products WHERE id = " . $productId . " AND active = 1");
    return mysqli_fetch_assoc($result);
}

// VULN 3: Double-quoted string with variable interpolation in SQL (SQL Injection)
function getProductsByCategory($conn, $categorySlug) {
    $pdo = new PDO("mysql:host=localhost;dbname=shopist_db", "shopist_user", "shopist_pass");
    // Variable interpolated directly into SQL string — PDO provides no benefit here
    $stmt = $pdo->query("SELECT p.* FROM products p JOIN categories c ON p.category_id = c.id WHERE c.slug = '$categorySlug' ORDER BY p.created_at DESC");
    return $stmt->fetchAll(PDO::FETCH_ASSOC);
}

// --- Route dispatcher ---
$action = $_GET['action'] ?? '';

if ($action === 'search') {
    $results = searchProducts($conn);
    header('Content-Type: application/json');
    echo json_encode($results);
}

if ($action === 'detail') {
    $product = getProductById($conn);
    header('Content-Type: application/json');
    echo json_encode($product);
}

if ($action === 'category') {
    $slug = $_GET['category'] ?? '';
    $products = getProductsByCategory($conn, $slug);
    header('Content-Type: application/json');
    echo json_encode($products);
}
